File: src/DSA/LinkedList/LRUCache.cs
using System.Collections.Generic;

namespace DSA.LinkedList
{
    // Basic LRU Cache Design
    // Most recently used item will be at front and least recently used item will be at tail
    public class LRUCache
    {
        private static readonly int DefaultCap = 100;
        private Node tail;
        private Node head;
        readonly Dictionary<int, Node> hashMap;
        private readonly int capacity;
        private int size;

        public LRUCache() : this(LRUCache.DefaultCap)
        {

        }

        public LRUCache(int cap)
        {
            capacity = cap;
            size = 0;
            head = tail = null;
            hashMap = new Dictionary<int, Node>();
        }

        private void DeleteTail()
        {
            if (tail == null || size == 0)
            {
                return;
            }

            hashMap.Remove(tail.Data);

            Node prev = tail.Prev;
            tail.Prev = null;
            tail.Next = null;
            prev.Next = null;
            tail = prev;
            size--;
        }

        void DeleteNode(Node node)
        {
            if (node == null)
                return;

            if (node.Prev != null)
            {
                node.Prev.Next = node.Next;
            }

            if (node.Next != null)
            {
                node.Next.Prev = node.Prev;
            }
            else
            {
                tail = node.Prev;
            }

            node.Prev = null;
            node.Next = null;
        }

        void AddNodeToHead(Node node)
        {
            if (node == null)
                return;

            if (head == null)
            {
                head = tail = node;
                return;
            }

            node.Next = head;
            head.Prev = node;
            head = node;
        }

        public int Get(int key)
        {
            if (hashMap.ContainsKey(key))
            {
                Node node = hashMap[key];

                if (node == head)
                {
                    return head.Data;
                }

                DeleteNode(node);

                AddNodeToHead(node);

                return node.Data;
            }

            return -1;
        }

        public void Set(int key, int value)
        {
            Node node;

            if (hashMap.ContainsKey(key))
            {
                node = hashMap[key];
                node.Data = value;
                if (node == head)
                {
                    return;
                }

                DeleteNode(node);

                AddNodeToHead(node);
            }
            else
            {
                if (size >= capacity)
                {
                    DeleteTail();
                }

                node = new Node() { Data = value };

                AddNodeToHead(node);
                hashMap.Add(key, node);
                size++;
            }
        }

        public void Clear()
        {
            head = null;
            tail = null;
            size = 0;
            hashMap.Clear();
        }

        public string Print()
        {
            string res = string.Empty;
            if (head == null)
            {
                res = "Cache is Empty";
                System.Console.WriteLine(res);
                return res;
            }
            else
            {
                Node curr = head;
                while (curr != null)
                {
                    res = res + curr.Data + " => ";
                    if (curr.Next == null)
                    {
                        res = res + "NULL";
                    }

                    curr = curr.Next;
                }
                System.Console.WriteLine(res);
                return res;
            }
        }
    }
}

File: java/src/linkedList/LRUCache.java
package linkedList;

import java.util.HashMap;

public class LRUCache {
    public class Node {
        public int key;
        public int data;
        public Node prev;
        public Node next;

        public Node(int key, int val) {
            this.key = key;
            this.data = val;
            this.prev = this.next = null;
        }
    }

    Node head, tail;
    HashMap<Integer, Node> map;

    int size, capacity;

    public LRUCache(int capacity) {
        this.capacity = capacity;
        this.map = new HashMap<Integer, Node>();
        this.head = null;
        this.tail = null;
        this.size = 0;
    }

    public int get(int key) {
        if (!map.containsKey(key)) {
            return -1;
        }

        Node n = this.map.get(key);

        this.deleteNode(n);

        this.addToFront(n);

        return n.data;
    }

    public void set(int key, int value) {

        Node n;
        if (map.containsKey(key)) {
            n = this.map.get(key);
            n.data = value;

            this.deleteNode(n);
        } else {
            n = new Node(key, value);

            this.map.put(key, n);

            if (isCacheFull()) {
                this.map.remove(this.tail.key);

                this.deleteNode(this.tail);

                this.size--;
            }

            this.size++;
        }

        this.addToFront(n);
    }

    private boolean isCacheFull() {
        return this.size == this.capacity;
    }

    private void addToFront(Node n) {
        if (this.head == null) {
            this.head = n;
            this.tail = n;
        } else {
            n.next = this.head;
            this.head.prev = n;
            this.head = n;
        }
    }

    private void deleteNode(Node n) {
        Node prev = n.prev, next = n.next;
        if (prev != null) {
            prev.next = next;
        }

        if (next != null) {
            next.prev = prev;
        }

        n.prev = null;
        n.next = null;

        if (prev == null) {
            this.head = next;
        }

        if (next == null) {
            this.tail = prev;
        }
    }
}